### DISCUSSION TITLE
2022-41: Rust Drop 踩坑分享

### DISCUSSION THREAD

**Discussion by giscus** [Xuanwo/blog#382] [Announcements]
# 2022-41: Rust Drop 踩坑分享

An infrastructure engineer, focused on distributed storage system

https://xuanwo.io/reports/2022-41/

**Comment by xxchan** [comment]
> 但是 let _ = abc; 的语义却更晦涩一些：不要将后面的表达式绑定为任何东西。它只是一个 match 表达式，本身并不会导致 drop，之所以我们观察到 drop 是因为它 match 的值本身就是临时的。

这句话不是很好理解，在 https://runrust.miraheze.org/wiki/Underscore#Underscores_in_let_bindings 看到了一个不错的解释:
1. `let _` work 实际上是因为 `let` acceps a *pattern*
2. `let _ = blabla;` 的实际效果等同于 `blabla;` 所以更精彩的事情是：`let _=x; let y=x;`

**Reply by Xuanwo** [reply]
我有读到过类似的解释，但是 `let _ = blabla;` 跟 `blabla;` 的语义还是不同的，所以我没有这样写上去。

有这样的[反例](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b57678e7a2da1a1b948c27135b5e8b47)：

```rust
struct Test(&'static str);

impl Drop for Test {
    fn drop(&mut self) {
        println!("Test with {} dropped", self.0)
    }
}

fn main() {
    let x = Test("x");
    {
        println!("into scope");
        let _ = x;
        println!("leave scope");
    }
    
    let y = Test("y");
    {
        println!("into scope");
        y;
        println!("leave scope");
    }
}
}
```

执行结果如下：

```txt
into scope
leave scope
into scope
Test with y dropped
leave scope
Test with x dropped
```

**Reply by xxchan** [reply]
好的，那更加震撼了
应该是和这个等价：
```
match blabla {
    _ => ()
}
```

**Reply by Xuanwo** [reply]
这个转换不错！我更新一把～

**Reply by xxchan** [reply]
其实不需要这么复杂的例子，这个就够了： 🤣
```
x;
let y=x;
```

**Comment by xxchan** [comment]
`let _` 既不是 no-op 也不是 drop，但是可以分类成
- `let _ = <variable>` 是 no-op
- `let _ = <expr>` 是 drop

🤔（`let _x = ...` 是 binding）

**Reply by lilydjwg** [reply]
实际上这个差异是被匹配项（scrutinee）的差异造成的。当被匹配项是值表达式（即 rvalue）时，值会被放入临时位置从而在语句结束时销毁。但被匹配项是位置表达式（即 lvalue）时，它并不会被放入额外的位置，若不被模式所引用或者移出，那么它不影响该值的生命周期（而 `_` 就是除了匹配一下啥也不做）。

这个 `let _` 一开始最令我困惑的地方在于，我以为它像 Python 那样是个变量名，结果它是个用于匹配的模式……
